### DISCUSSION TITLE
[Eloquent Model] initialize traits after filling model to dynamically add attributes

### DISCUSSION THREAD

**Discussion by juliomotol** [laravel/framework#36141] [Ideas]
Is there a way to where we could initialize traits for a model after filling its attributes?

https://github.com/laravel/framework/blob/73955837fed7ce38dda460c1734995165fb6c923/src/Illuminate/Database/Eloquent/Model.php#L166-L175

The reasoning for this is I want my trait to initialize an attribute that depends on another column.

```php
trait FooTrait
{
    protected function initializeFooTrait()
    {
        $this->attribute['dynamicly_loaded_mutator'] = 'some derived value from another attribute';
    }
}
```

During `initializeFooTrait()`, the model's attributes isn't filled yet, thus throwing errors.

Maybe add something like `finalizeTraits()` where you define `finilizeFooTrait()` in your trait.

I'd be very happy to make a PR for this <3

**Comment by austrowGC** [comment]
This seems like it could be handled using the 'retrieved' Model event, but I could be wrong. I'm looking into using it for a similar use case.

**Reply by juliomotol** [reply]
Hmm, I'll give that a try, I'll keep you posted on that!

**Reply by austrowGC** [reply]
Per the documentation on [Eloquent Events](https://laravel.com/docs/8.x/eloquent#events), adding this to your model should accomplish what you described:
```php
    protected static function booted()
    {
        static::retrieved( function ($model) {
            $model->attribute['dynamically_loaded_mutator'] = 'some derived value from another attribute';
        });
    }
```

You could also create an Event subclass to take the place of the anonymous function, if you prefer that organization or need to use it elsewhere.

**Reply by juliomotol** [reply]
I've been skimming the `Model` class but the `$attribute` property is protected and it wont be mutable within the closure, unless you use `$this->attribute` but I'm very skeptical in doing that.

Maybe having a `saving` events that clears the added attributes from the `retrieved`. But it feels like running all around the places. 🤔

**Reply by austrowGC** [reply]
I'm confused about what you want to do, then. As far as I know, the protected `$attributes` member on a model is only supposed to represent the default values for fields. In the retrieved callback, you can directly set the value that you need to `$model->attribute`. The value of `$model` passed to that closure has data from the underlying table at that point and you can invoke whatever logic you need--I was just copying what you had used as a placeholder in your initial post.

**Reply by juliomotol** [reply]
Disregard what said about the `saving` event part, looking at the `save()` method of `Model`, it seems that this would work. I'll be doing my initial test later today.

**Reply by devfrey** [reply]
> I've been skimming the `Model` class but the `$attribute` **property is protected and it wont be mutable within the closure**, unless you use `$this->attribute` but I'm very skeptical in doing that.

@juliomotol You can access `$attributes` within your closure just fine, since the closure is defined within the same class.

**Reply by yob-yob** [reply]
One `Cons` for this is that it's using the `booted()` method.. which would can cause a collision from other traits..

**Comment by UncertaintyP** [comment]
Because `$this->attribute` must be a property of your trait or you specific model (because Laravel does **NOT** have that property) you can set its visibility as you like and it must be available at startup. What is the error you are receiving?

**Comment by trane77** [comment]
Hi, 
I'm probably facing the same issue as @juliomotol. 

To give you a little more context:

- you have many Model classes using the trait
- the trait will modify the model, e.g. by adding an attribute
- you want to exclude the trait initialization on a specific condition, which depends on the value of another attribute

Right now you can't, as attributes are filled *after* trait(s) initialization.

This is very likely a corner case, but having the data available, would allow a more granular use of Traits.


### Edit - a solution to this problem 

1. Trait class

```php
trait XYZTrait
{
    public bool $hasXYZTrait = true;  // will use this later to easily check if model uses the trait

    // ... trait content

    public function isExcludedType(): bool
    {
      
        if ($this instanceof YourModelToCheck) {
            return $this->propertyToCheck === 'valueThatExcludesTrait';
        }

        return $false;
    }

}
```

2. Example use 

```php
$modelHasXYZTrait = property_exists($model, 'hasXYZTrait');

$modelHasActiveTrait = $modelHasXYZTrait && $visitModel->isExcludedType() === false;
```

The `property_exists()` call is used to easily check if the model uses the trait. Correct me if I'm wrong, but php doesn't provide an easy solution to that.

**Reply by juliomotol** [reply]
Hi @funder7, I've forgotten what I was trying to solve and solution for this. But based on your case, you might be able to use [models events](https://laravel.com/docs/11.x/eloquent#events) as mentioned by @austrowGC

```php
trait FooTrait
{
    public static function bootFooTrait()
    {
        self::retrieved(function (static $model) {
            if ($condition) {
                $model->dynamic_attribute = $value;
            }
        });
    }
}
```

If this attribute is derived from another and you want this to also be updated every time that has been updated, you may use a [mutator](https://laravel.com/docs/11.x/eloquent-mutators#defining-a-mutator).

```php
protected function foo(): Attribute
{
    return Attribute::set(function ($value) {
        return with(
            ['foo' => $value], // You'll need to passthrough the dependency attribute so that It can also be saved.
            function ($items) {
                if ($condition) {
                    $items['dynamic_attribute'] => $someValue;
                }
            }
        );
    });
}
```

> Note: Without the mutator, the attribute will never be filled when creating a new instance since it will only be loaded when retrieved.

Although from what I can remember about the origin of this discussion, I didn't want this attribute to be saved almost acting like an accessor. If that's also the case for you, you can use  the `saving` event in the `boot` to `unset()` the added attribute

```php
self::saving(function (static $model) {
    unset($model->dynamic_attribute);
});
```

Again I can't recall why I would go through the hassle of doing that (as you can see in some of my responses in the other comments), but a simpler approach would be to just use an accessor for this.

```php
protected function dynamicAttribute(): Attribute
{
    return Attribute::get(function (){
        if $condition
            ? $value
            : null
    });
}
```
